Squeezer/src/main/java/uk/org/ngo/squeezer/service/SqueezeService.java

```diff
@@ -775,23 +822,149 @@ private void downloadSong(@NonNull Uri url, String title, @NonNull Uri serverUrl
+                                            @Override
+                                            public void onScanCompleted(String path, final Uri uri) {
+                                                if (!Uri.EMPTY.equals(downloadEntry.albumArtUrl)) {
+                                                    // It seems that onScanCompleted is called off thread
+                                                    // (though I can find no documentation for it)
+                                                    // so we make this run on the main thread, as the ImageFetcher
+                                                    // may need it (if it creates a new cache)
+                                                    mMainThreadHandler.post(new Runnable() {
+                                                        @Override
+                                                        public void run() {
+                                                            downloadAlbumArt(uri, downloadEntry.albumArtUrl);
+                                                        }
+                                                    });
+                                                }
+                                            }
+                                        }
+                                );
+                            } catch (IOException e) {
+                                Util.crashlyticsLogException(e);
+                            }
+                            break;
+                        default:
+                            Util.crashlyticsLog(Log.ERROR, TAG, "Unsuccessful download " + format(status, reason, title, url, local_url));
+                            break;
+                    }
+                } else {
+                    Util.crashlyticsLog(Log.ERROR, TAG, "Download database does not have an entry for " + format(status, reason, title, url, local_url));
+                }
+            //} else {
+            // Download complete events may still come in, even after DownloadManager.remove is
+            // called, so don't log this
+            //Logger.logError(TAG, "Download manager does not have an entry for " + id);
+            }
+        } finally {
+            cursor.close();
+        }
+    }
+
+    // TODO hack
+    // This is a hack, as it relies on internals of the AOSP MediaProvider class, and the below
+    // non-public URI.
+    // We mimic the behavior of the media provider when it creates an album thumb, so a media player
+    // will receive the location of our downloaded album art when queriyng
+    // MediaStore.Audio.Albums.ALBUM_ART.
+    private static final Uri sArtworkUri = Uri.parse("content://media/external/audio/albumart");
+    private void downloadAlbumArt(Uri songUri, Uri albumArtUrl) {
+        final Cursor songCursor = getContentResolver().query(songUri, new String[] { MediaStore.Audio.Media.ALBUM_ID }, null, null, null);
+        if (songCursor != null) {
```

Comment:
Currently what Java gives us for safe release of resources (try with resources requires API level 19), is the construct:
```
    Cursor cursor = ...
    try {
        while (cursor.moveToNext()) {
            doSomething();
        }
    } finally {
        cursor.close();
    }
```

I think this should be used. It makes sure the resource is released even on failure (almost anything can throw), and also handle early returns.

I rearranged the code a bit to reduce the complexity.

In SqueezeService.handleDownloadComplete:
* Early returns for missing download manger entry, missing download database entry and unsuccessful download. 
* Convert anonymous OnScanCompletedListener implementation to inner.

In SqueezeService.downloadAlbumArt:
* Return early if songCursor is null, like you suggested.
* Convert anonymous ImageWorkerCallback implementation to inner.

On a side note using 2 spaces for indentation would also make the code more readable :-)